### PR TITLE
feat(module:date-picker): quarter selection of date picker

### PR DESCRIPTION
**Pull request by m310851010**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] Application (the showcase website) / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #7818, #7380, #3124

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: components/i18n/date-helper.service.ts [hunk 1]
```diff
@@ -120,4 +119,14 @@ export class DateHelperByDatePipe extends DateHelperService {
     const parser = new ɵNgTimeParser(formatStr, this.i18n.getLocaleId());
     return parser.toDate(text);
   }
+
+  private getQuarter(date: Date): number {
+    return Math.floor((date.getMonth() + 3) / 3);
+  }
```
Comment (HyperLife1119) [inline components/i18n/date-helper.service.ts:125]: 使用 date-fns 的 getQuarter 替代

File: components/i18n/date-helper.service.ts [hunk 2]
```diff
@@ -120,4 +119,14 @@ export class DateHelperByDatePipe extends DateHelperService {
...
+
+  private replaceQuarter(dateStr: string, date: Date): string {
+    const quarter = this.getQuarter(date).toString();
+    const record: Record<string, string> = { Q: quarter, QQ: `0${quarter}`, QQQ: `Q${quarter}` };
+    return dateStr.replace(/Q+(?![^\[]*])/g, match => record[match] ?? quarter).replace(/\[([^[\]]*Q[^[\]]*)]/g, '$1');
+  }
```
Comment (HyperLife1119) [inline components/i18n/date-helper.service.ts:131]: 简化正则表达式，并添加一些注释：

```ts
// Q Pattern format compatible with date-fns (quarter).
return dateStr
  // Match Q+ outside of brackets, then replace it with the specified quarterly format
  .replace(/Q+(?![^\[]*])/g, match => record[match] ?? quarter)
  // Match the Q+ surrounded by bracket, then remove bracket.
  .replace(/\[(Q+)]/g, '$1');
```

File: components/date-picker/demo/date-render.ts [hunk 3]
```diff
@@ -7,9 +7,13 @@ import { Component } from '@angular/core';
     <nz-range-picker [nzDateRender]="tplRender"></nz-range-picker>
 
     <ng-template #tplRender let-current>
-      <div class="ant-picker-cell-inner" [class.border]="current.getDate() === 1">
-        {{ current.getDate() }}
-      </div>
+      <div class="ant-picker-cell-inner" [class.border]="current.getDate() === 1"> {{ current.getDate() }} </div>
```
Comment (HyperLife1119) [inline components/date-picker/demo/date-render.ts:10]:
```suggestion
      <div class="ant-picker-cell-inner" [class.border]="current.getDate() === 1">{{ current.getDate() }}</div>
```

File: components/i18n/date-helper.service.ts [hunk 4]
```diff
@@ -118,4 +119,17 @@ export class DateHelperByDatePipe extends DateHelperService {
...
+    const record: Record<string, string> = { Q: quarter, QQ: `0${quarter}`, QQQ: `Q${quarter}` };
+    // Q Pattern format compatible with date-fns (quarter).
+    return (
+      dateStr
+        // Match Q+ outside of brackets, then replace it with the specified quarterly format
+        .replace(/Q+(?![^\[]*])/g, match => record[match] ?? quarter)
+        // Match the Q+ surrounded by bracket, then remove bracket.
+        .replace(/\[(Q+)]/g, '$1')
+    );
+  }
```
Comment (HyperLife1119) [inline components/i18n/date-helper.service.ts:134]: 可以为新增的 Q 格式添加单元测试 :)

**Review by HyperLife1119 — APPROVED** [review]
LGTM

**Comment by HyperLife1119** [comment]
@Nicoss54 Can you review this PR together? :)

**Comment by Nicoss54** [comment]
> @Nicoss54 Can you review this PR together? :)

of course i will review it today or tomorrow morning :)

**Review by Nicoss54 — APPROVED** [review]
LGTM

File: components/i18n/languages/zh_TW.ts [hunk 5]
```diff
@@ -98,6 +98,7 @@ export default {
       monthPlaceholder: '請選擇月份',
       weekPlaceholder: '請選擇周',
       rangeYearPlaceholder: ['開始年份', '結束年份'],
+      rangeQuarterPlaceholder: ['開始季度', '開始季度'],
```
Comment (Laffery) [inline components/i18n/languages/zh_TW.ts:101]:
```diff
- rangeQuarterPlaceholder: ['開始季度', '開始季度'],
+ rangeQuarterPlaceholder: ['開始季度', '結束季度'],
```

File: components/i18n/languages/zh_HK.ts [hunk 6]
```diff
@@ -53,6 +53,7 @@ export default {
       monthPlaceholder: '請選擇月份',
       weekPlaceholder: '請選擇周',
       rangeYearPlaceholder: ['開始年份', '結束年份'],
+      rangeQuarterPlaceholder: ['開始季度', '開始季度'],
```
Comment (Laffery) [inline components/i18n/languages/zh_HK.ts:56]:
```diff
- rangeQuarterPlaceholder: ['開始季度', '開始季度'],
+ rangeQuarterPlaceholder: ['開始季度', '結束季度'],
```

**Review by Laffery — CHANGES_REQUESTED** [review]
Hi there 👋, thank your very much for contributing such fancy feature to ng-zorro, there are some issues should be fixed :)

**Review by Laffery — APPROVED** [review]
LGTM

**Pull request merged**